text/0249-v2-experiments.md

```diff
@@ -408,3 +484,240 @@ Disadvantages:
+| Go    | ?           | ?                                |                    | ?                           |
+
+**NPM**
+
+In NPM-based languages (JavaScript and TypeScript), the package name appears
+in source code (in the `require()` statement). If we changed the library
+name, we wouldn't be able to use a library that uses stable code (it would
+contain `require('aws-cdk-lib')`) in an application that uses experimental
+code (the `package.json` would contain: `{ dependencies: {
+"aws-cdk-lib-experiments": "..." }}`).
+
+NPM 6.9.0 (released May 2019) introduces a feature that helps with this:
+[package
+aliases](https://github.com/npm/rfcs/blob/latest/implemented/0001-package-aliases.md)
+which allows installing a package under a different name, so we could install
+`aws-cdk-lib-experimental` under the name `aws-cdk-lib`, and no `require()` statements
+would have to be changed.
+
+An application trying to use experimental will do this:
+
+```json
+{
+  "dependencies": {
+      "aws-cdk-lib": "npm:aws-cdk-lib-experiments@1.60.0",
+      "some-construct-lib-that-uses-stable": "^1.2.3"
+  }
+}
+```
+
+An experimental library would declare this:
+
+```json
+{
+  "peerDependencies": {
+      "aws-cdk-lib": "1.60.0"
+      // Note: ideally you'd like to express "aws-cdk-lib-experiments": "1.60.0" here, but
+      // if you do that "npm ls" will complain about an unmet peer dependency
+  },
+  "devDependencies": {
+      "aws-cdk-lib": "npm:aws-cdk-lib-experiments@1.60.0"
+  }
+}
+```
+
+This will all work as intended and put code in the right places,
+but we lose the tool-checked dependency on experimental; from NPM's
+point of view the library depends on stable CDK, and it's only
+notes in the README that can inform the consumer otherwise.
+
+**Maven**
+
+No need to change the sources when using a stable library with an
+experimental app, as the namespaces and class names will be the same.
+
+Using a stable library in an experimental app can be done in two ways:
+
+* The library uses the `provided` dependency flavor (similar to `peerDependencies` in NPM).
+  Requires the app to bring the right version of CDK.
+* The app that wants to use the *experimental* build uses dependency exclusions
+  to remove the dependency on the *stable* build (looks like this is necessary for
+  every stable library individually)
+
+https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html#
+
+**pip**
+
+No need to change the sources when using a stable library with an
+experimental app, as the namespaces and class names will be the same.
+
+`pip` transitively fetches from the library's `setup.py`, and there's no way
+to override that.
+
+Our only option around that is to have stable libraries not declare any CDK
+dependency in `setup.py`, but then we also lose the ability for libraries to
+specify their dependency version, which is not really acceptable.
```

Comment:
Is it really such a big deal? I feel like declaring the CDK dependency version is just meant to provide a different error message when you violate it? So is it really such a big deal that it would fail at runtime?